distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmOptionsParser.java

```diff
@@ -115,6 +117,24 @@ public void accept(final int lineNumber, final String line) {
         }
     }
 
+    static List<String> substitutePlaceholders(final List<String> jvmOptions, Map<String, String> substitutions) {
+        final Map<String, String> placeholderSubstitutions =
+            substitutions.entrySet().stream().collect(Collectors.toMap(e -> "${" + e.getKey() + "}", Map.Entry::getValue));
+        final List<String> substitutedJvmOptions = new ArrayList<>(jvmOptions.size());
+        for (final String jvmOption : jvmOptions) {
+            if (jvmOption.matches(".*\\$\\{[^}]+\\}.*")) {
+                String actualJvmOption = jvmOption;
+                for (final Map.Entry<String, String> placeholderSubstitution : placeholderSubstitutions.entrySet()) {
+                   actualJvmOption = actualJvmOption.replace(placeholderSubstitution.getKey(), placeholderSubstitution.getValue());
+                }
+                substitutedJvmOptions.add(actualJvmOption);
+            } else {
+                substitutedJvmOptions.add(jvmOption);
+            }
+        }
+        return substitutedJvmOptions;
+    }
+
```

Comment:
This method can be simplified with use of `mathce.appendReplacement` method:
```java
private static final Pattern SUBSTITUTION_PATTERN = Pattern.compile("\\$\\{([^}]+)}");

static List<String> substitutePlaceholders(final List<String> jvmOptions, Map<String, String> substitutions) {
    return jvmOptions.stream().map(jvmOption -> {
        StringBuilder sb = new StringBuilder();
        Matcher matcher = SUBSTITUTION_PATTERN.matcher(jvmOption);
        while (matcher.find()) {
            matcher.appendReplacement(sb, substitutions.get(matcher.group(1)));
        }
        matcher.appendTail(sb);
        return sb.toString();
    }).collect(Collectors.toList());
}
```
Or made faster (~2x) when you replace regex with ```indexOf```:
```java
static List<String> substitutePlaceholders(final List<String> jvmOptions, Map<String, String> substitutions) {
    final Map<String, String> placeholderSubstitutions =
            substitutions.entrySet().stream().collect(Collectors.toMap(e -> "${" + e.getKey() + "}", Map.Entry::getValue));
    return jvmOptions.stream().map(jvmOption -> {
        String actualJvmOption = jvmOption;
        int start = jvmOption.indexOf("${");
        if (start >= 0 && jvmOption.indexOf('}', start) > 0) {
            for (final Map.Entry<String, String> placeholderSubstitution : placeholderSubstitutions.entrySet()) {
                actualJvmOption = actualJvmOption.replace(placeholderSubstitution.getKey(), placeholderSubstitution.getValue());
            }
        }
        return actualJvmOption;
    }).collect(Collectors.toList());
}
```

Reply:
Thanks for the suggestion @probakowski, I pushed it.